### DISCUSSION TITLE
Regarding Adding Image in background

### DISCUSSION THREAD

**Discussion by ZubairImran** [ShapeCrawler/ShapeCrawler#777] [Q&A]
Hi Team,
I am using below code snippet to add images in slide. 
 foreach (ISlide slide in presentation.Slides)
                {
                    using (var stream = new FileStream(stImagePath, FileMode.Open))
                    {
                       slide.Shapes.AddPicture(stream);
                    }
                }
i have below doubts
1. is it possilble to add the image in background(SendBack option in pptx). currently with the above code , the image is getting added on top and its preventing further edit on the slide
2.  while adding the image and saving the presentation increases size of the presentation, is there a way save the images in compressed format(without increasing the presentation ) and open it without any problem

Thanks

**Comment by ZubairImran** [comment]
@ashahabov, any hint on the above would be great help.


Thanks

**Reply by AdamShakhabov** [reply]
Hello @ZubairImran,

### 1. Background Update
If I'm currently understood, the first question is about updating the slide background, which you can do with the following code

```C#
var pre = new Presentation("pres.pptx");
var image = File.ReadAllBytes("image.jpg");
var background = pre.Slide(1).Background;

background.Update(image);

pre.Save();
```

Let me know if your question is about another.

### 2. Image compression
The image compression is out of ShapeCrawler. You can achieve this using another library, for example, [SkiaSharp](https://github.com/mono/SkiaSharp) and then use the compressed image in ShapeCrawler

```C#
using var inputStream = File.OpenRead("input.jpg");
using var originalBitmap = SKBitmap.Decode(inputStream);

// Scale down the resolution by 50%
int newWidth = originalBitmap.Width / 2;
int newHeight = originalBitmap.Height / 2;
var samplingOptions = new SKSamplingOptions();
using var resizedBitmap = originalBitmap.Resize(new SKImageInfo(newWidth, newHeight), samplingOptions);

using var image = SKImage.FromBitmap(resizedBitmap);
using var outputStream = File.OpenWrite("output.jpg");
image.Encode(SKEncodedImageFormat.Jpeg, 90) // Compression quality set to 90 (you can adjust this)
        .SaveTo(outputStream);
```

**Reply by ZubairImran** [reply]
@ashahabov , Thanks for the detailed answer. I will try these and update you. Regarding the first question about background, i could able to add the image but the image which i have added is preventing the text update behind that image. However i will try the approach mentioned and get back.

**Reply by ZubairImran** [reply]
@ashahabov ,i have tried the answer for the question 1 and below my observation.
* when i try to set the image as backaground , it was adding two or three images in the background. But my doubt is different , sorry that i couldn't've explained it earlier.   I am using the below code to add the picture in the slide

<code>
         var presentation = new Presentation(pptxFilePath);
				// Iterate over each slide in the presentation
                foreach (ISlide slide in presentation.Slides)
                {
                    var imagePath = imgpath;

                    using (var stream = new FileStream(imagePath, FileMode.Open))
                    {
                       slide.Shapes.AddPicture(stream);

                    }
	          }
</code>
   The above is working fine and image is placed in each slides left corner which was fine. the above code internally place the images at the position (100,100) and internally converted in to EMUs. My Question is , the picture added like this can be sent to background . Now the added images showing on top due to which i couldnt able update text under it. Is it possible to image back. in PPTX application if we right click on the image there will be option "SendBack" in context menu.

**Reply by AdamShakhabov** [reply]
Oh, I got it!

No, we didn't have this feature, but now I've implemented this method `SendToBack()` which you can test using the latest beta version.

```Shell
dotnet add package ShapeCrawler --version 0.55.2-beta.202411211024 --source https://www.myget.org/F/shape/api/v3/index.json
```

```C#
var pre = new Presentation("pre.pptx");
var shapes = pre.Slide(1).Shapes;
var picture = shapes.GetByName<IPicture>("Picture 1");

picture.SendToBack();
```

**Reply by ZubairImran** [reply]
Sure @ashahabov . I will test and get back to you. Thanks for this implementation.

**Reply by ZubairImran** [reply]
`PM> dotnet add CheckSend package ShapeCrawler --version 0.55.2-beta.202411211024 --source https://www.myget.org/F/shape/api/v3/index.json
  Determining projects to restore...
  Writing C:\Users\zimran\AppData\Local\Temp\tmp42pbz2.tmp
info : X.509 certificate chain validation will use the default trust store selected by .NET for code signing.
info : X.509 certificate chain validation will use the default trust store selected by .NET for timestamping.
info : Adding PackageReference for package 'ShapeCrawler' into project 'C:\SendBack_PPT\CheckSend\CheckSend\CheckSend.csproj'.
error: Error while adding package 'ShapeCrawler' to project 'C:\SendBack_PPT\CheckSend\CheckSend\CheckSend.csproj'. The project does not support adding package references through the add package command.`

**Reply by ZubairImran** [reply]
@ashahabov , Seems i am doing some mistake, is it possible to share NuGet source or source(i will build it) to test the above. sorry for the inconvenience.

**Reply by AdamShakhabov** [reply]
[ShapeCrawler.0.56.1-beta.202411232040.nupkg.zip](https://github.com/user-attachments/files/17901866/ShapeCrawler.0.56.1-beta.202411232040.nupkg.zip)

**Reply by AdamShakhabov** [reply]
`dotnet add ...` doesn't work also on my side😕 

It seems we need to use this [**option**](https://github.com/ShapeCrawler/ShapeCrawler?tab=readme-ov-file#prerelease-version) to be able to add prerelease version.

**Reply by ZubairImran** [reply]
@ashahabov , thanks for the prompt reply, i will try with provided option.

**Reply by ZubairImran** [reply]
@ashahabov , I have tried with package(attached nuget.zip) you sent. I have got the below exception(System contains no matchine element) , Am i missing anything.
```
System.InvalidOperationException
  HResult=0x80131509
  Message=Sequence contains no matching element
  Source=System.Core
  StackTrace:
   at System.Linq.Enumerable.First[TSource](IEnumerable`1 source, Func`2 predicate)
   at ShapeCrawler.ShapeCollection.Shapes.GetByName(String name)
   at ShapeCrawler.ShapeCollection.Shapes.GetByName[T](String name)
   at ShapeCrawler.ShapeCollection.SlideShapes.GetByName[T](String name)
   at CheckSend.Program.Main(String[] args) in C:\SendBack_PPT\CheckSend\CheckSend\Program.cs:line 16
```

**Reply by ZubairImran** [reply]
The below code i have used with argument "Picture" or "Picture 1"  both giving above error
`
        static void Main(string[] args)
        {
            var pre = new Presentation("C:\\USERS\\PUBLIC\\sendback.pptx");
            var shapes = pre.Slide(1).Shapes;
            var picture = shapes.GetByName<IPicture>("Picture");

            picture.SendToBack();
            pre.Save();
        }`

**Reply by AdamShakhabov** [reply]
Please double-check that your picture shape has the name "Picture" or "Picture 1":

![picture](https://github.com/user-attachments/assets/8a9ab062-39ab-4659-8010-2fb1afc1d3e7)

**Reply by ZubairImran** [reply]
hi @ashahabov , thanks for the above hint. Now its working fine. Thanks so much. When it will be available in main stream.

**Reply by AdamShakhabov** [reply]
The next release ETA is 29.11.

**Reply by ZubairImran** [reply]
Thanks @ashahabov.
